ccl/multiregionccl: TestColdStartLatency failed

ccl/multiregionccl.TestColdStartLatency [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8401354?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8401354?buildTab=artifacts#/) on master @ [2861a4ee05fac1cc13f87bf00f368a65ea8700dc](https://github.com/cockroachdb/cockroach/commits/2861a4ee05fac1cc13f87bf00f368a65ea8700dc):


```
=== RUN   TestColdStartLatency
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/e558fc8050776f4c54ea39ba371b49da/logTestColdStartLatency1194377700
    test_log_scope.go:79: use -show-logs to present logs inline
    cold_start_latency_test.go:170: 0 alter database system configure zone discard;
    cold_start_latency_test.go:170: 1 alter database system set primary region "us-east1";
    cold_start_latency_test.go:170: 2 alter database system add region "us-west1";
    cold_start_latency_test.go:170: 3 alter database system add region "europe-west1"
    cold_start_latency_test.go:170: 4 ALTER TENANT ALL SET CLUSTER SETTING sql.zone_configs.allow_for_secondary_tenant.enabled = true
    cold_start_latency_test.go:170: 5 ALTER TENANT ALL SET CLUSTER SETTING sql.multi_region.allow_abstractions_for_secondary_tenants.enabled = true
    cold_start_latency_test.go:170: 6 alter range meta configure zone using constraints = '{"+region=us-east1": 1, "+region=us-west1": 1, "+region=europe-west1": 1}';
    cold_start_latency_test.go:170: 0 alter database system configure zone discard;
    cold_start_latency_test.go:170: 1 alter database system set primary region "us-east1";
    cold_start_latency_test.go:170: 2 alter database system add region "us-west1";
    cold_start_latency_test.go:170: 3 alter database system add region "europe-west1"
    cold_start_latency_test.go:170: 4 SELECT crdb_internal.unsafe_optimize_system_database()
    cold_start_latency_test.go:363: pre running test to allocate instance IDs
    cold_start_latency_test.go:344: done 8 region=us-west1,az=c 1.570294909s
    cold_start_latency_test.go:344: done 3 region=us-east1,az=a 571.906605ms
    cold_start_latency_test.go:344: done 0 region=europe-west1,az=a 1.409910818s
    cold_start_latency_test.go:344: done 1 region=europe-west1,az=b 499.363892ms
    cold_start_latency_test.go:344: done 2 region=europe-west1,az=c 535.971721ms
    cold_start_latency_test.go:344: done 5 region=us-east1,az=c 524.720458ms
    cold_start_latency_test.go:344: done 4 region=us-east1,az=b 456.400384ms
    cold_start_latency_test.go:344: done 6 region=us-west1,az=a 522.674191ms
    cold_start_latency_test.go:344: done 7 region=us-west1,az=b 543.887202ms
    cold_start_latency_test.go:364: result [region=europe-west1,az=a region=europe-west1,az=b region=europe-west1,az=c region=us-east1,az=a region=us-east1,az=b region=us-east1,az=c region=us-west1,az=a region=us-west1,az=b region=us-west1,az=c] [1.409910818s 499.363892ms 535.971721ms 571.906605ms 456.400384ms 524.720458ms 522.674191ms 543.887202ms 1.570294909s]
    cold_start_latency_test.go:365: running test with no connectivity from sql pods to remote regions
    cold_start_latency_test.go:344: done 8 region=us-west1,az=c 551.240653ms
    cold_start_latency_test.go:344: done 0 region=europe-west1,az=a 486.696041ms
    cold_start_latency_test.go:344: done 4 region=us-east1,az=b 542.193272ms
    cold_start_latency_test.go:344: done 1 region=europe-west1,az=b 592.083249ms
    cold_start_latency_test.go:344: done 2 region=europe-west1,az=c 512.0611ms
    cold_start_latency_test.go:344: done 3 region=us-east1,az=a 534.555646ms
panic: IsNull on unset EncDatum

goroutine 91676 [running]:
github.com/cockroachdb/cockroach/pkg/sql/rowenc.(*EncDatum).IsNull(0xc02b9fd3e0?)
	github.com/cockroachdb/cockroach/pkg/sql/rowenc/encoded_datum.go:221 +0x1c5
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sketchInfo).addRow(0xc023dbd330, {0x6cb2290, 0xc029b7bb90}, {0xc02b9fd320, 0x6, 0xc0eb0227a5a10e15?}, {0xc021d80270, 0xe, 0x1?}, 0xc02bd359e0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:505 +0x91
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*samplerProcessor).mainLoop(0xc02dec6d00, {0x6cb2290, 0xc029b7bb90})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:320 +0x1d92
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*samplerProcessor).Run(0xc02dec6d00, {0x6cb21e8?, 0xc025086700?})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:229 +0x9a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).StartInternal.func1(0xc004ccdfb8?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:447 +0x48
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).StartInternal
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:446 +0x395
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/multiregion
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestColdStartLatency.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23647

ccl/multiregionccl.TestColdStartLatency [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8461860?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8461860?buildTab=artifacts#/) on master @ [2ad8df3df3272110705984efc32f1453631ce602](https://github.com/cockroachdb/cockroach/commits/2ad8df3df3272110705984efc32f1453631ce602):


```
=== RUN   TestColdStartLatency
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/e558fc8050776f4c54ea39ba371b49da/logTestColdStartLatency1388645235
    test_log_scope.go:79: use -show-logs to present logs inline
    cold_start_latency_test.go:173: 0 alter database system configure zone discard;
    cold_start_latency_test.go:173: 1 alter database system set primary region "us-east1";
    cold_start_latency_test.go:173: 2 alter database system add region "us-west1";
    cold_start_latency_test.go:173: 3 alter database system add region "europe-west1"
    cold_start_latency_test.go:173: 4 ALTER TENANT ALL SET CLUSTER SETTING sql.zone_configs.allow_for_secondary_tenant.enabled = true
    cold_start_latency_test.go:173: 5 ALTER TENANT ALL SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false
    cold_start_latency_test.go:173: 6 ALTER TENANT ALL SET CLUSTER SETTING sql.multi_region.allow_abstractions_for_secondary_tenants.enabled = true
    cold_start_latency_test.go:173: 7 alter range meta configure zone using constraints = '{"+region=us-east1": 1, "+region=us-west1": 1, "+region=europe-west1": 1}';
    cold_start_latency_test.go:173: 0 alter database system configure zone discard;
    cold_start_latency_test.go:173: 1 alter database system set primary region "us-east1";
    cold_start_latency_test.go:173: 2 alter database system add region "us-west1";
    cold_start_latency_test.go:173: 3 alter database system add region "europe-west1"
    cold_start_latency_test.go:173: 4 SELECT crdb_internal.unsafe_optimize_system_database()
panic: IsNull on unset EncDatum

goroutine 105737 [running]:
github.com/cockroachdb/cockroach/pkg/sql/rowenc.(*EncDatum).IsNull(0xc0108561e0?)
	github.com/cockroachdb/cockroach/pkg/sql/rowenc/encoded_datum.go:221 +0x1c5
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sketchInfo).addRow(0xc02dd6f730, {0x6d06fd0, 0xc0097f2720}, {0xc010856120, 0x6, 0xc0eca8a18b5a946d?}, {0xc022971ee0, 0xe, 0x2?}, 0xc014e539e0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:505 +0x91
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*samplerProcessor).mainLoop(0xc01f05e000, {0x6d06fd0, 0xc0097f2720})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:320 +0x1d92
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*samplerProcessor).Run(0xc01f05e000, {0x6d06fd0?, 0xc029a7a300?})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:229 +0x9a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).StartInternal.func1(0xc026cc0a98?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:447 +0x48
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).StartInternal
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:446 +0x395
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestColdStartLatency.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/multiregionccl.TestColdStartLatency [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8462005?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8462005?buildTab=artifacts#/) on master @ [9caf75810ff7dbfa42d16b746c21c86bfa0c6324](https://github.com/cockroachdb/cockroach/commits/9caf75810ff7dbfa42d16b746c21c86bfa0c6324):


```
=== RUN   TestColdStartLatency
I230126 08:00:02.116790 1 (gostd) rand.go:199  [T1] 1  random seed: -5826563925970078173
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/201913953f2c683da40d46e0bffd06a3/logTestColdStartLatency358393331
    test_log_scope.go:79: use -show-logs to present logs inline
    cold_start_latency_test.go:173: 0 alter database system configure zone discard;
    cold_start_latency_test.go:173: 1 alter database system set primary region "us-east1";
    cold_start_latency_test.go:173: 2 alter database system add region "us-west1";
    cold_start_latency_test.go:173: 3 alter database system add region "europe-west1"
    cold_start_latency_test.go:173: 4 ALTER TENANT ALL SET CLUSTER SETTING sql.zone_configs.allow_for_secondary_tenant.enabled = true
    cold_start_latency_test.go:173: 5 ALTER TENANT ALL SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false
    cold_start_latency_test.go:173: 6 ALTER TENANT ALL SET CLUSTER SETTING sql.multi_region.allow_abstractions_for_secondary_tenants.enabled = true
    cold_start_latency_test.go:173: 7 alter range meta configure zone using constraints = '{"+region=us-east1": 1, "+region=us-west1": 1, "+region=europe-west1": 1}';
    cold_start_latency_test.go:173: 0 alter database system configure zone discard;
    cold_start_latency_test.go:173: 1 alter database system set primary region "us-east1";
    cold_start_latency_test.go:173: 2 alter database system add region "us-west1";
    cold_start_latency_test.go:173: 3 alter database system add region "europe-west1"
    cold_start_latency_test.go:173: 4 SELECT crdb_internal.unsafe_optimize_system_database()
    cold_start_latency_test.go:366: pre running test to allocate instance IDs
    cold_start_latency_test.go:347: done 8 region=us-west1,az=c 1.360146142s
    cold_start_latency_test.go:347: done 3 region=us-east1,az=a 315.135036ms
    cold_start_latency_test.go:347: done 0 region=europe-west1,az=a 2.496853963s
    cold_start_latency_test.go:347: done 1 region=europe-west1,az=b 388.491325ms
    cold_start_latency_test.go:347: done 2 region=europe-west1,az=c 255.593314ms
    cold_start_latency_test.go:347: done 5 region=us-east1,az=c 296.088848ms
    cold_start_latency_test.go:347: done 4 region=us-east1,az=b 307.045453ms
    cold_start_latency_test.go:347: done 6 region=us-west1,az=a 351.151789ms
    cold_start_latency_test.go:347: done 7 region=us-west1,az=b 467.733941ms
    cold_start_latency_test.go:367: result [region=europe-west1,az=a region=europe-west1,az=b region=europe-west1,az=c region=us-east1,az=a region=us-east1,az=b region=us-east1,az=c region=us-west1,az=a region=us-west1,az=b region=us-west1,az=c] [2.496853963s 388.491325ms 255.593314ms 315.135036ms 307.045453ms 296.088848ms 351.151789ms 467.733941ms 1.360146142s]
    cold_start_latency_test.go:368: running test with no connectivity from sql pods to remote regions
    cold_start_latency_test.go:347: done 8 region=us-west1,az=c 396.563226ms
    cold_start_latency_test.go:347: done 1 region=europe-west1,az=b 600.206094ms
    cold_start_latency_test.go:347: done 0 region=europe-west1,az=a 381.460368ms
    cold_start_latency_test.go:347: done 4 region=us-east1,az=b 504.987514ms
panic: IsNull on unset EncDatum

goroutine 97886 [running]:
github.com/cockroachdb/cockroach/pkg/sql/rowenc.(*EncDatum).IsNull(0x4012994420?)
	github.com/cockroachdb/cockroach/pkg/sql/rowenc/encoded_datum.go:221 +0x1e4
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sketchInfo).addRow(0x401f99db30, {0x6516d30, 0x400eb9d0e0}, {0x4012994360, 0x6, 0x26ecee290f49802e?}, {0x40177d3ba0, 0xe, 0x402fd8cc00?}, 0x4022e1d9d0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:505 +0x80
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*samplerProcessor).mainLoop(0x4024b66d00, {0x6516d30, 0x400eb9d0e0})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:320 +0x155c
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*samplerProcessor).Run(0x4024b66d00, {0x6516d30?, 0x400eb9cf90?})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/sampler.go:229 +0x74
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).StartInternal.func1(0x400c3faad0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:447 +0x54
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).StartInternal
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:446 +0x2c8
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestColdStartLatency.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/multiregionccl/cold_start_latency_test.go
+++ b/pkg/ccl/multiregionccl/cold_start_latency_test.go
@@ -49,7 +49,6 @@ func TestColdStartLatency(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	skip.UnderRace(t, "too slow")
 	skip.UnderStress(t, "too slow")
-	skip.WithIssue(t, 95644)
 	defer envutil.TestSetEnv(t, "COCKROACH_MR_SYSTEM_DATABASE", "1")()
 
 	// We'll need to make some per-node args to assign the different
@@ -160,9 +159,6 @@ func TestColdStartLatency(t *testing.T) {
 		if !isTenant {
 			stmts = append(stmts,
 				"ALTER TENANT ALL SET CLUSTER SETTING sql.zone_configs.allow_for_secondary_tenant.enabled = true",
-				// Disable automatic stats because it gets upset about the way we change the column
-				// structure.
-				"ALTER TENANT ALL SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false",
 				"ALTER TENANT ALL SET CLUSTER SETTING sql.multi_region.allow_abstractions_for_secondary_tenants.enabled = true",
 				`alter range meta configure zone using constraints = '{"+region=us-east1": 1, "+region=us-west1": 1, "+region=europe-west1": 1}';`)
 		} else {

--- a/pkg/sql/sqlinstance/instancestorage/BUILD.bazel
+++ b/pkg/sql/sqlinstance/instancestorage/BUILD.bazel
@@ -28,7 +28,6 @@ go_library(
         "//pkg/sql/catalog/systemschema",
         "//pkg/sql/enum",
         "//pkg/sql/rowenc/valueside",
-        "//pkg/sql/sem/catid",
         "//pkg/sql/sem/tree",
         "//pkg/sql/sqlinstance",
         "//pkg/sql/sqlliveness",
@@ -49,6 +48,7 @@ go_library(
 go_test(
     name = "instancestorage_test",
     srcs = [
+        "helpers_test.go",
         "instancereader_test.go",
         "instancestorage_internal_test.go",
         "instancestorage_test.go",
@@ -67,8 +67,8 @@ go_test(
         "//pkg/security/securitytest",
         "//pkg/server",
         "//pkg/settings/cluster",
-        "//pkg/sql/catalog/descpb",
         "//pkg/sql/catalog/descs",
+        "//pkg/sql/catalog/desctestutils",
         "//pkg/sql/catalog/systemschema",
         "//pkg/sql/enum",
         "//pkg/sql/sqlinstance",
@@ -76,6 +76,7 @@ go_test(
         "//pkg/sql/sqlliveness/slstorage",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
+        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
         "//pkg/util/encoding",
@@ -89,6 +90,7 @@ go_test(
         "//pkg/util/timeutil",
         "//pkg/util/uuid",
         "@com_github_cockroachdb_errors//:errors",
+        "@com_github_stretchr_testify//assert",
         "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/sql/sqlinstance/instancestorage/helpers_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/helpers_test.go
@@ -1,0 +1,31 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package instancestorage
+
+import (
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+)
+
+// GetTableSQLForDatabase is a testing helper to construct the appropriate
+// table schema for a given database taking into consideration whether we're
+// configured for the MR schema.
+func GetTableSQLForDatabase(dbName string) string {
+	schema := systemschema.SQLInstancesTableSchema
+	if systemschema.TestSupportMultiRegion() {
+		schema = systemschema.MrSQLInstancesTableSchema
+	}
+	schema = strings.Replace(schema,
+		`CREATE TABLE system.sql_instances`,
+		`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+	return schema
+}

--- a/pkg/sql/sqlinstance/instancestorage/instancereader.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancereader.go
@@ -20,8 +20,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/sem/catid"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
@@ -59,7 +59,7 @@ func NewTestingReader(
 	slReader sqlliveness.Reader,
 	f *rangefeed.Factory,
 	codec keys.SQLCodec,
-	tableID descpb.ID,
+	table catalog.TableDescriptor,
 	clock *hlc.Clock,
 	stopper *stop.Stopper,
 ) *Reader {
@@ -69,7 +69,7 @@ func NewTestingReader(
 		f:               f,
 		codec:           codec,
 		clock:           clock,
-		rowcodec:        makeRowCodec(codec, tableID),
+		rowcodec:        makeRowCodec(codec, table),
 		initialScanDone: make(chan struct{}),
 		stopper:         stopper,
 	}
@@ -86,7 +86,7 @@ func NewReader(
 	clock *hlc.Clock,
 	stopper *stop.Stopper,
 ) *Reader {
-	return NewTestingReader(storage, slReader, f, codec, keys.SQLInstancesTableID, clock, stopper)
+	return NewTestingReader(storage, slReader, f, codec, systemschema.SQLInstancesTable(), clock, stopper)
 }
 
 // Start initializes the rangefeed for the Reader. The rangefeed will run until
@@ -144,17 +144,16 @@ func makeInstanceInfos(rows []instancerow) []sqlinstance.InstanceInfo {
 // GetAllInstancesUsingTxn reads all instances using the given transaction and returns
 // live instances only.
 func (r *Reader) GetAllInstancesUsingTxn(
-	ctx context.Context, codec keys.SQLCodec, tableID catid.DescID, txn *kv.Txn,
+	ctx context.Context, txn *kv.Txn,
 ) ([]sqlinstance.InstanceInfo, error) {
-	instancesTablePrefix := codec.TablePrefix(uint32(tableID))
+	instancesTablePrefix := r.rowcodec.codec.TablePrefix(uint32(r.rowcodec.tableID))
 	rows, err := txn.Scan(ctx, instancesTablePrefix, instancesTablePrefix.PrefixEnd(), 0 /* maxRows */)
 	if err != nil {
 		return nil, err
 	}
-	rowcodec := makeRowCodec(codec, tableID)
 	decodedRows := make([]instancerow, 0, len(rows))
 	for _, row := range rows {
-		decodedRow, err := rowcodec.decodeRow(row.Key, row.Value)
+		decodedRow, err := r.rowcodec.decodeRow(row.Key, row.Value)
 		if err != nil {
 			return nil, err
 		}

--- a/pkg/sql/sqlinstance/instancestorage/instancereader_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancereader_test.go
@@ -12,7 +12,6 @@ package instancestorage_test
 
 import (
 	"context"
-	"strings"
 	"testing"
 	"time"
 
@@ -21,7 +20,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/enum"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
@@ -54,14 +53,12 @@ func TestReader(t *testing.T) {
 	) {
 		dbName := t.Name()
 		tDB.Exec(t, `CREATE DATABASE "`+dbName+`"`)
-		schema := strings.Replace(systemschema.SQLInstancesTableSchema,
-			`CREATE TABLE system.sql_instances`,
-			`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+		schema := instancestorage.GetTableSQLForDatabase(dbName)
 		tDB.Exec(t, schema)
-		tableID := getTableID(t, tDB, dbName, "sql_instances")
+		table := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), dbName, "sql_instances")
 		slStorage := slstorage.NewFakeStorage()
-		storage := instancestorage.NewTestingStorage(s.DB(), keys.SystemSQLCodec, tableID, slStorage, s.ClusterSettings())
-		reader := instancestorage.NewTestingReader(storage, slStorage, s.RangeFeedFactory().(*rangefeed.Factory), keys.SystemSQLCodec, tableID, s.Clock(), s.Stopper())
+		storage := instancestorage.NewTestingStorage(s.DB(), keys.SystemSQLCodec, table, slStorage, s.ClusterSettings())
+		reader := instancestorage.NewTestingReader(storage, slStorage, s.RangeFeedFactory().(*rangefeed.Factory), keys.SystemSQLCodec, table, s.Clock(), s.Stopper())
 		return storage, slStorage, s.Clock(), reader
 	}
 
@@ -123,7 +120,7 @@ func TestReader(t *testing.T) {
 					return errors.Newf("expected instance address %s != actual instance address %s", rpcAddr, instanceInfo.InstanceRPCAddr)
 				}
 				if sqlAddr != instanceInfo.InstanceSQLAddr {
-					return errors.Newf("expected instance address %s != actual instance address %s", sqlAddr, instanceInfo.InstanceSQLAddr)
+					return errors.Newf("expected instance sql address %s != actual sql instance address %s", sqlAddr, instanceInfo.InstanceSQLAddr)
 				}
 				if !locality.Equals(instanceInfo.Locality) {
 					return errors.Newf("expected instance locality %s != actual instance locality %s", locality, instanceInfo.Locality)
@@ -189,7 +186,7 @@ func TestReader(t *testing.T) {
 			var instancesUsingTxn []sqlinstance.InstanceInfo
 			if err := s.DB().Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 				var err error
-				instancesUsingTxn, err = reader.GetAllInstancesUsingTxn(ctx, s.Codec(), getTableID(t, tDB, t.Name(), "sql_instances"), txn)
+				instancesUsingTxn, err = reader.GetAllInstancesUsingTxn(ctx, txn)
 				return err
 			}); err != nil {
 				return nil, err

--- a/pkg/sql/sqlinstance/instancestorage/instancestorage.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancestorage.go
@@ -27,8 +27,9 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/enum"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
@@ -115,13 +116,13 @@ func (r *instancerow) isAvailable() bool {
 func NewTestingStorage(
 	db *kv.DB,
 	codec keys.SQLCodec,
-	sqlInstancesTableID descpb.ID,
+	table catalog.TableDescriptor,
 	slReader sqlliveness.Reader,
 	settings *cluster.Settings,
 ) *Storage {
 	s := &Storage{
 		db:       db,
-		rowcodec: makeRowCodec(codec, sqlInstancesTableID),
+		rowcodec: makeRowCodec(codec, table),
 		slReader: slReader,
 		settings: settings,
 	}
@@ -132,7 +133,7 @@ func NewTestingStorage(
 func NewStorage(
 	db *kv.DB, codec keys.SQLCodec, slReader sqlliveness.Reader, settings *cluster.Settings,
 ) *Storage {
-	return NewTestingStorage(db, codec, keys.SQLInstancesTableID, slReader, settings)
+	return NewTestingStorage(db, codec, systemschema.SQLInstancesTable(), slReader, settings)
 }
 
 // CreateNodeInstance claims a unique instance identifier for the SQL pod, and

--- a/pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go
@@ -14,7 +14,6 @@ import (
 	"context"
 	gosql "database/sql"
 	"sort"
-	"strings"
 	"testing"
 	"time"
 
@@ -23,8 +22,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/enum"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
@@ -65,7 +63,7 @@ func TestGetAvailableInstanceIDForRegion(t *testing.T) {
 	}
 
 	t.Run("no rows", func(t *testing.T) {
-		stopper, storage, _, _ := setup(t, sqlDB, kvDB, s.ClusterSettings())
+		stopper, storage, _, _ := setup(t, sqlDB, kvDB, s.Codec(), s.ClusterSettings())
 		defer stopper.Stop(ctx)
 
 		id, err := getAvailableInstanceID(storage, nil)
@@ -75,7 +73,7 @@ func TestGetAvailableInstanceIDForRegion(t *testing.T) {
 
 	t.Run("preallocated rows", func(t *testing.T) {
 		const expiration = time.Minute
-		stopper, storage, slStorage, clock := setup(t, sqlDB, kvDB, s.ClusterSettings())
+		stopper, storage, slStorage, clock := setup(t, sqlDB, kvDB, s.Codec(), s.ClusterSettings())
 		defer stopper.Stop(ctx)
 
 		// Pre-allocate four instances.
@@ -300,7 +298,7 @@ func TestReclaimAndGenerateInstanceRows(t *testing.T) {
 	regions := [][]byte{enum.One}
 
 	t.Run("nothing preallocated", func(t *testing.T) {
-		stopper, storage, _, clock := setup(t, sqlDB, kvDB, s.ClusterSettings())
+		stopper, storage, _, clock := setup(t, sqlDB, kvDB, s.Codec(), s.ClusterSettings())
 		defer stopper.Stop(ctx)
 
 		sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
@@ -320,7 +318,7 @@ func TestReclaimAndGenerateInstanceRows(t *testing.T) {
 	})
 
 	t.Run("with some preallocated", func(t *testing.T) {
-		stopper, storage, slStorage, clock := setup(t, sqlDB, kvDB, s.ClusterSettings())
+		stopper, storage, slStorage, clock := setup(t, sqlDB, kvDB, s.Codec(), s.ClusterSettings())
 		defer stopper.Stop(ctx)
 
 		sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
@@ -421,20 +419,6 @@ func TestReclaimAndGenerateInstanceRows(t *testing.T) {
 	})
 }
 
-func getTableID(
-	t *testing.T, db *sqlutils.SQLRunner, dbName, tableName string,
-) (tableID descpb.ID) {
-	t.Helper()
-	db.QueryRow(t, `
- select u.id
-  from system.namespace t
-  join system.namespace u
-  on t.id = u."parentID"
-  where t.name = $1 and u.name = $2`,
-		dbName, tableName).Scan(&tableID)
-	return tableID
-}
-
 func sortInstancesForTest(instances []sqlinstance.InstanceInfo) {
 	sort.SliceStable(instances, func(idx1, idx2 int) bool {
 		addr1, addr2 := instances[idx1].InstanceRPCAddr, instances[idx2].InstanceRPCAddr
@@ -456,20 +440,18 @@ func sortInstancesForTest(instances []sqlinstance.InstanceInfo) {
 }
 
 func setup(
-	t *testing.T, sqlDB *gosql.DB, kvDB *kv.DB, settings *cluster.Settings,
+	t *testing.T, sqlDB *gosql.DB, kvDB *kv.DB, codec keys.SQLCodec, settings *cluster.Settings,
 ) (*stop.Stopper, *Storage, *slstorage.FakeStorage, *hlc.Clock) {
 	dbName := t.Name()
 	tDB := sqlutils.MakeSQLRunner(sqlDB)
 	tDB.Exec(t, `CREATE DATABASE "`+dbName+`"`)
-	schema := strings.Replace(systemschema.SQLInstancesTableSchema,
-		`CREATE TABLE system.sql_instances`,
-		`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+	schema := GetTableSQLForDatabase(dbName)
 	tDB.Exec(t, schema)
-	tableID := getTableID(t, tDB, dbName, "sql_instances")
+	table := desctestutils.TestingGetPublicTableDescriptor(kvDB, codec, dbName, "sql_instances")
 	clock := hlc.NewClock(timeutil.NewTestTimeSource(), base.DefaultMaxClockOffset)
 	stopper := stop.NewStopper()
 	slStorage := slstorage.NewFakeStorage()
-	storage := NewTestingStorage(kvDB, keys.SystemSQLCodec, tableID, slStorage, settings)
+	storage := NewTestingStorage(kvDB, keys.SystemSQLCodec, table, slStorage, settings)
 	return stopper, storage, slStorage, clock
 }
 

--- a/pkg/sql/sqlinstance/instancestorage/instancestorage_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancestorage_test.go
@@ -16,7 +16,6 @@ import (
 	"fmt"
 	"math/rand"
 	"sort"
-	"strings"
 	"sync"
 	"testing"
 	"time"
@@ -24,8 +23,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/enum"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance"
@@ -34,6 +33,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -43,6 +43,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
@@ -70,15 +71,13 @@ func TestStorage(t *testing.T) {
 	) {
 		dbName := t.Name()
 		tDB.Exec(t, `CREATE DATABASE "`+dbName+`"`)
-		schema := strings.Replace(systemschema.SQLInstancesTableSchema,
-			`CREATE TABLE system.sql_instances`,
-			`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+		schema := instancestorage.GetTableSQLForDatabase(dbName)
 		tDB.Exec(t, schema)
-		tableID := getTableID(t, tDB, dbName, "sql_instances")
+		table := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), dbName, "sql_instances")
 		clock := hlc.NewClock(timeutil.NewTestTimeSource(), base.DefaultMaxClockOffset)
 		stopper := stop.NewStopper()
 		slStorage := slstorage.NewFakeStorage()
-		storage := instancestorage.NewTestingStorage(kvDB, keys.SystemSQLCodec, tableID, slStorage, s.ClusterSettings())
+		storage := instancestorage.NewTestingStorage(kvDB, keys.SystemSQLCodec, table, slStorage, s.ClusterSettings())
 		return stopper, storage, slStorage, clock
 	}
 
@@ -278,15 +277,13 @@ func TestSQLAccess(t *testing.T) {
 	tDB := sqlutils.MakeSQLRunner(sqlDB)
 	dbName := t.Name()
 	tDB.Exec(t, `CREATE DATABASE "`+dbName+`"`)
-	schema := strings.Replace(systemschema.SQLInstancesTableSchema,
-		`CREATE TABLE system.sql_instances`,
-		`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+	schema := instancestorage.GetTableSQLForDatabase(dbName)
 	tDB.Exec(t, schema)
-	tableID := getTableID(t, tDB, dbName, "sql_instances")
+	table := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), dbName, "sql_instances")
 	stopper := stop.NewStopper()
 	defer stopper.Stop(ctx)
 	storage := instancestorage.NewTestingStorage(
-		kvDB, keys.SystemSQLCodec, tableID, slstorage.NewFakeStorage(), s.ClusterSettings())
+		kvDB, keys.SystemSQLCodec, table, slstorage.NewFakeStorage(), s.ClusterSettings())
 	const (
 		tierStr         = "region=test1,zone=test2"
 		expiration      = time.Minute
@@ -315,7 +312,10 @@ func TestSQLAccess(t *testing.T) {
 	var parsedAddr gosql.NullString
 	var parsedSqlAddr gosql.NullString
 	var parsedLocality gosql.NullString
-	rows.Next()
+	if !assert.True(t, rows.Next()) {
+		require.NoError(t, rows.Err())
+		require.NoError(t, rows.Close())
+	}
 	err = rows.Scan(&parsedInstanceID, &parsedAddr, &parsedSqlAddr, &parsedSessionID, &parsedLocality)
 	require.NoError(t, err)
 	require.Equal(t, instance.InstanceID, parsedInstanceID)
@@ -342,6 +342,10 @@ func TestSQLAccess(t *testing.T) {
 func TestRefreshSession(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
+
+	if systemschema.TestSupportMultiRegion() {
+		skip.IgnoreLintf(t, "this test gets run twice and is not set up for use with the MR schema")
+	}
 
 	ctx := context.Background()
 	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{Locality: roachpb.Locality{Tiers: []roachpb.Tier{{Key: "abc", Value: "xyz"}}}})
@@ -398,15 +402,13 @@ func TestConcurrentCreateAndRelease(t *testing.T) {
 	tDB := sqlutils.MakeSQLRunner(sqlDB)
 	dbName := t.Name()
 	tDB.Exec(t, `CREATE DATABASE "`+dbName+`"`)
-	schema := strings.Replace(systemschema.SQLInstancesTableSchema,
-		`CREATE TABLE system.sql_instances`,
-		`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+	schema := instancestorage.GetTableSQLForDatabase(dbName)
 	tDB.Exec(t, schema)
-	tableID := getTableID(t, tDB, dbName, "sql_instances")
+	table := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), dbName, "sql_instances")
 	stopper := stop.NewStopper()
 	slStorage := slstorage.NewFakeStorage()
 	defer stopper.Stop(ctx)
-	storage := instancestorage.NewTestingStorage(kvDB, keys.SystemSQLCodec, tableID, slStorage, s.ClusterSettings())
+	storage := instancestorage.NewTestingStorage(kvDB, keys.SystemSQLCodec, table, slStorage, s.ClusterSettings())
 	instancestorage.PreallocatedCount.Override(ctx, &s.ClusterSettings().SV, 1)
 
 	const (
@@ -547,11 +549,9 @@ func TestReclaimLoop(t *testing.T) {
 	tDB := sqlutils.MakeSQLRunner(sqlDB)
 	dbName := t.Name()
 	tDB.Exec(t, `CREATE DATABASE "`+dbName+`"`)
-	schema := strings.Replace(systemschema.SQLInstancesTableSchema,
-		`CREATE TABLE system.sql_instances`,
-		`CREATE TABLE "`+dbName+`".sql_instances`, 1)
+	schema := instancestorage.GetTableSQLForDatabase(dbName)
 	tDB.Exec(t, schema)
-	tableID := getTableID(t, tDB, dbName, "sql_instances")
+	tableID := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), dbName, "sql_instances")
 	slStorage := slstorage.NewFakeStorage()
 	storage := instancestorage.NewTestingStorage(kvDB, keys.SystemSQLCodec, tableID, slStorage, s.ClusterSettings())
 	storage.TestingKnobs.JitteredIntervalFn = func(d time.Duration) time.Duration {
@@ -675,20 +675,6 @@ func TestReclaimLoop(t *testing.T) {
 			require.Empty(t, instance.Locality)
 		}
 	}
-}
-
-func getTableID(
-	t *testing.T, db *sqlutils.SQLRunner, dbName, tableName string,
-) (tableID descpb.ID) {
-	t.Helper()
-	db.QueryRow(t, `
- select u.id
-  from system.namespace t
-  join system.namespace u
-  on t.id = u."parentID"
-  where t.name = $1 and u.name = $2`,
-		dbName, tableName).Scan(&tableID)
-	return tableID
 }
 
 func sortInstances(instances []sqlinstance.InstanceInfo) {

--- a/pkg/sql/sqlinstance/instancestorage/main_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/main_test.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 )
 
@@ -26,7 +27,22 @@ func TestMain(m *testing.M) {
 	randutil.SeedForTests()
 	serverutils.InitTestServerFactory(server.TestServerFactory)
 	serverutils.InitTestClusterFactory(testcluster.TestClusterFactory)
+
+	// We care about running all of these tests both in the "regular" config
+	// and in the "COCKROACH_MR_SYSTEM_DATABASE" config. This is certainly
+	// hacky, but go supports it, begrudgingly (see golang.org/issues/23129).
+	if code := m.Run(); code != 0 {
+		os.Exit(code)
+	}
+	defer envutil.TestSetEnv(
+		fakeTestingTB{}, "COCKROACH_MR_SYSTEM_DATABASE", "t",
+	)()
 	os.Exit(m.Run())
 }
 
 //go:generate ../../../util/leaktest/add-leaktest.sh *_test.go
+
+// fakeTestingTB is used to access envutil.TestSetEnv in TestMain.
+type fakeTestingTB struct{ testing.TB }
+
+func (fakeTestingTB) Helper() {}

--- a/pkg/sql/sqlinstance/instancestorage/row_codec.go
+++ b/pkg/sql/sqlinstance/instancestorage/row_codec.go
@@ -18,7 +18,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/enum"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc/valueside"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
@@ -48,10 +47,34 @@ type keyCodec interface {
 // rowCodec encodes/decodes rows from the sql_instances table.
 type rowCodec struct {
 	keyCodec
-	codec   keys.SQLCodec
-	columns []catalog.Column
-	decoder valueside.Decoder
-	tableID descpb.ID
+	codec               keys.SQLCodec
+	columns             []catalog.Column
+	decoder             valueside.Decoder
+	tableID             descpb.ID
+	valueColumnIDs      [numValueColumns]descpb.ColumnID
+	valueColumnOrdinals [numValueColumns]int
+}
+
+type valueColumnIdx int
+
+const numValueColumns = 4
+
+const (
+	addrColumnIdx valueColumnIdx = iota
+	sessionIDColumnIdx
+	localityColumnIdx
+	sqlAddrColumnIdx
+
+	// Ensure we have the right number of value columns.
+	_ uint = iota - numValueColumns
+	_ uint = numValueColumns - (iota - 1)
+)
+
+var valueColumnNames = [numValueColumns]string{
+	addrColumnIdx:      "addr",
+	sessionIDColumnIdx: "session_id",
+	localityColumnIdx:  "locality",
+	sqlAddrColumnIdx:   "sql_addr",
 }
 
 // rbrKeyCodec is used by the regional by row compatible sql_instances index format.
@@ -127,27 +150,31 @@ func (c *rbtKeyCodec) decodeKey(key roachpb.Key) (region []byte, id base.SQLInst
 }
 
 // MakeRowCodec makes a new rowCodec for the sql_instances table.
-func makeRowCodec(codec keys.SQLCodec, tableID descpb.ID) rowCodec {
-	columns := systemschema.SQLInstancesTable().PublicColumns()
-
+func makeRowCodec(codec keys.SQLCodec, table catalog.TableDescriptor) rowCodec {
+	columns := table.PublicColumns()
 	var key keyCodec
-	if systemschema.TestSupportMultiRegion() {
+	if catalog.FindColumnByName(table, "crdb_region") != nil {
 		key = &rbrKeyCodec{
-			indexPrefix: codec.IndexPrefix(uint32(tableID), 2),
+			indexPrefix: codec.IndexPrefix(uint32(table.GetID()), uint32(table.GetPrimaryIndexID())),
 		}
 	} else {
 		key = &rbtKeyCodec{
-			indexPrefix: codec.IndexPrefix(uint32(tableID), 1),
+			indexPrefix: codec.IndexPrefix(uint32(table.GetID()), uint32(table.GetPrimaryIndexID())),
 		}
 	}
-
-	return rowCodec{
+	rc := rowCodec{
 		keyCodec: key,
 		codec:    codec,
 		columns:  columns,
 		decoder:  valueside.MakeDecoder(columns),
-		tableID:  tableID,
+		tableID:  table.GetID(),
 	}
+	for i := 0; i < numValueColumns; i++ {
+		col := catalog.FindColumnByName(table, valueColumnNames[i])
+		rc.valueColumnIDs[i] = col.GetID()
+		rc.valueColumnOrdinals[i] = col.Ordinal()
+	}
+	return rc
 }
 
 // decodeRow converts the key and value into an instancerow. value may be nil
@@ -180,51 +207,44 @@ func (d *rowCodec) encodeValue(
 	rpcAddr string, sqlAddr string, sessionID sqlliveness.SessionID, locality roachpb.Locality,
 ) (*roachpb.Value, error) {
 	var valueBuf []byte
-
-	rpcAddrDatum := tree.DNull
-	if rpcAddr != "" {
-		rpcAddrDatum = tree.NewDString(rpcAddr)
+	columnsToEncode := [numValueColumns]func() tree.Datum{
+		addrColumnIdx: func() tree.Datum {
+			if rpcAddr == "" {
+				return tree.DNull
+			}
+			return tree.NewDString(rpcAddr)
+		},
+		sessionIDColumnIdx: func() tree.Datum {
+			if len(sessionID) == 0 {
+				return tree.DNull
+			}
+			return tree.NewDBytes(tree.DBytes(sessionID.UnsafeBytes()))
+		},
+		localityColumnIdx: func() tree.Datum {
+			if len(locality.Tiers) == 0 {
+				return tree.DNull
+			}
+			builder := json.NewObjectBuilder(1)
+			builder.Add("Tiers", json.FromString(locality.String()))
+			return tree.NewDJSON(builder.Build())
+		},
+		sqlAddrColumnIdx: func() tree.Datum {
+			if sqlAddr == "" {
+				return tree.DNull
+			}
+			return tree.NewDString(sqlAddr)
+		},
 	}
-	sqlAddrDatum := tree.DNull
-	if sqlAddr != "" {
-		sqlAddrDatum = tree.NewDString(sqlAddr)
-	}
-
-	// Ordering of the values below needs to remain stable for backwards
-	// compatibility. New values should be appended to the end.
-	valueBuf, err := valueside.Encode(
-		[]byte(nil), valueside.MakeColumnIDDelta(0, d.columns[1].GetID()), rpcAddrDatum, []byte(nil))
-	if err != nil {
-		return nil, err
-	}
-
-	sessionDatum := tree.DNull
-	if len(sessionID) > 0 {
-		sessionDatum = tree.NewDBytes(tree.DBytes(sessionID.UnsafeBytes()))
-	}
-	sessionColDiff := valueside.MakeColumnIDDelta(d.columns[1].GetID(), d.columns[2].GetID())
-	valueBuf, err = valueside.Encode(valueBuf, sessionColDiff, sessionDatum, []byte(nil))
-	if err != nil {
-		return nil, err
-	}
-
-	// Preserve the ordering of locality.Tiers, even though we convert it to json.
-	localityDatum := tree.DNull
-	if len(locality.Tiers) > 0 {
-		builder := json.NewObjectBuilder(1)
-		builder.Add("Tiers", json.FromString(locality.String()))
-		localityDatum = tree.NewDJSON(builder.Build())
-	}
-	localityColDiff := valueside.MakeColumnIDDelta(d.columns[2].GetID(), d.columns[3].GetID())
-	valueBuf, err = valueside.Encode(valueBuf, localityColDiff, localityDatum, []byte(nil))
-	if err != nil {
-		return nil, err
-	}
-
-	valueBuf, err = valueside.Encode(valueBuf,
-		valueside.MakeColumnIDDelta(d.columns[3].GetID(), d.columns[4].GetID()), sqlAddrDatum, []byte(nil))
-	if err != nil {
-		return nil, err
+	for i, f := range columnsToEncode {
+		var err error
+		var prev descpb.ColumnID
+		if i > 0 {
+			prev = d.valueColumnIDs[i-1]
+		}
+		delta := valueside.MakeColumnIDDelta(prev, d.valueColumnIDs[i])
+		if valueBuf, err = valueside.Encode(valueBuf, delta, f(), nil); err != nil {
+			return nil, err
+		}
 	}
 
 	v := &roachpb.Value{}
@@ -251,52 +271,68 @@ func (d *rowCodec) decodeValue(
 	timestamp hlc.Timestamp,
 	_ error,
 ) {
-	// The rest of the columns are stored as a family.
+	// The rest of the columns are stored as a single family.
 	bytes, err := value.GetTuple()
 	if err != nil {
 		return "", "", "", roachpb.Locality{}, hlc.Timestamp{}, err
 	}
-
 	datums, err := d.decoder.Decode(&tree.DatumAlloc{}, bytes)
 	if err != nil {
 		return "", "", "", roachpb.Locality{}, hlc.Timestamp{}, err
 	}
-
-	if addrVal := datums[1]; addrVal != tree.DNull {
-		rpcAddr = string(tree.MustBeDString(addrVal))
-	}
-	if len(datums) == 5 {
-		if sqlAddrVal := datums[4]; sqlAddrVal != tree.DNull {
-			sqlAddr = string(tree.MustBeDString(sqlAddrVal))
-		} else {
-			// Backwards compatible with single-address version.
-			sqlAddr = rpcAddr
-		}
-	} else {
-		// Backwards compatible with single-address version.
-		sqlAddr = rpcAddr
-	}
-	if sessionIDVal := datums[2]; sessionIDVal != tree.DNull {
-		sessionID = sqlliveness.SessionID(tree.MustBeDBytes(sessionIDVal))
-	}
-	if localityVal := datums[3]; localityVal != tree.DNull {
-		localityJ := tree.MustBeDJSON(localityVal)
-		v, err := localityJ.FetchValKey("Tiers")
-		if err != nil {
-			return "", "", "", roachpb.Locality{}, hlc.Timestamp{}, errors.Wrap(err, "failed to find Tiers attribute in locality")
-		}
-		if v != nil {
-			vStr, err := v.AsText()
-			if err != nil {
-				return "", "", "", roachpb.Locality{}, hlc.Timestamp{}, err
+	for i, f := range [numValueColumns]func(datum tree.Datum) error{
+		addrColumnIdx: func(datum tree.Datum) error {
+			if datum != tree.DNull {
+				rpcAddr = string(tree.MustBeDString(datum))
 			}
-			if len(*vStr) > 0 {
-				if err := locality.Set(*vStr); err != nil {
-					return "", "", "", roachpb.Locality{}, hlc.Timestamp{}, err
+			return nil
+		},
+		sessionIDColumnIdx: func(datum tree.Datum) error {
+			if datum != tree.DNull {
+				sessionID = sqlliveness.SessionID(tree.MustBeDBytes(datum))
+			}
+			return nil
+		},
+		localityColumnIdx: func(datum tree.Datum) error {
+			if datum == tree.DNull {
+				return nil
+			}
+			localityJ := tree.MustBeDJSON(datum)
+			v, err := localityJ.FetchValKey("Tiers")
+			if err != nil {
+				return errors.Wrap(err, "failed to find Tiers attribute in locality")
+			}
+			if v != nil {
+				vStr, err := v.AsText()
+				if err != nil {
+					return err
+				}
+				if len(*vStr) > 0 {
+					if err := locality.Set(*vStr); err != nil {
+						return err
+					}
 				}
 			}
+			return nil
+		},
+		sqlAddrColumnIdx: func(datum tree.Datum) error {
+			if datum != tree.DNull {
+				sqlAddr = string(tree.MustBeDString(datum))
+			} else {
+				sqlAddr = rpcAddr
+			}
+			return nil
+		},
+	} {
+		ord := d.valueColumnOrdinals[i]
+		// Deal with the fact that new columns may not yet have been added.
+		datum := tree.DNull
+		if ord < len(datums) {
+			datum = datums[ord]
+		}
+		if err := f(datum); err != nil {
+			return "", "", "", roachpb.Locality{}, hlc.Timestamp{}, err
 		}
 	}
-
 	return rpcAddr, sqlAddr, sessionID, locality, value.Timestamp, nil
 }

--- a/pkg/sql/sqlinstance/instancestorage/row_codec_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/row_codec_test.go
@@ -26,7 +26,7 @@ import (
 )
 
 const (
-	tableID  = 42
+	tableID  = 46
 	tenantID = 1337
 )
 
@@ -40,11 +40,11 @@ func TestRowCodec(t *testing.T) {
 
 	t.Run("RegionalByRow", func(t *testing.T) {
 		defer envutil.TestSetEnv(t, "COCKROACH_MR_SYSTEM_DATABASE", "1")()
-		testEncoder(t, makeRowCodec(codec, tableID), tenantID)
+		testEncoder(t, makeRowCodec(codec, systemschema.SQLInstancesTable()), tenantID)
 	})
 	t.Run("RegionalByTable", func(t *testing.T) {
 		defer envutil.TestSetEnv(t, "COCKROACH_MR_SYSTEM_DATABASE", "0")()
-		testEncoder(t, makeRowCodec(codec, tableID), tenantID)
+		testEncoder(t, makeRowCodec(codec, systemschema.SQLInstancesTable()), tenantID)
 	})
 }
 

